gui/main.cpp

```diff
@@ -24,20 +31,61 @@ int main(int argc, char *argv[]) {
   parser.addOption(startupOnlyOption);
   parser.process(app);
 
-  auto generate_data = []() -> std::vector<std::tuple<float, float>> {
+  GuiStateContainer state_container(
+      /*history_window=*/std::chrono::milliseconds(30000));
+  auto startup_time = state_container.GetStartupTime();
+
+  std::unique_ptr<ConnectedDevice> device =
+      std::make_unique<FakeConnectedDevice>(
+          [&](const GuiStatus &gui_status,
+              ControllerStatus *controller_status) {
+            // For now, completely fake the data and ignore gui_status.
+            (void)gui_status;
+
+            controller_status->uptime_ms =
+                std::chrono::duration_cast<std::chrono::milliseconds>(
+                    std::chrono::steady_clock::now() - startup_time)
+                    .count();
+            auto *sensors = &controller_status->sensor_readings;
+            sensors->pressure_cm_h2o =
+                sin(controller_status->uptime_ms * 0.001);
+            sensors->volume_ml = 0;
+            sensors->flow_ml_per_min = 0;
+          });
+
+  // TODO: Bind the readable aspects of state_container to UI elements
+  // TODO: Bind the writable aspects of state_container to parameters set in the
+  // UI
+  PeriodicClosure maintain_history(std::chrono::milliseconds(10), [&]() {
+    ControllerStatus controller_status;
+    device->ExchangeState(state_container.GetGuiStatus(), &controller_status);
+    state_container.AppendHistory(controller_status);
+  });
+  maintain_history.Start();
```

Comment:
I'm following the API of the Google-internal one; AFAIK it is motivated by the fact that often you might have a base class declare a PeriodicClosure that invokes a virtual function; in that case, starting it during the base class constructor is unsafe, instead it's up to the subclass to Start() it when the subclass is constructed (in a method, rather than in a constructor). We don't have such a case here yet, but seems reasonable to do the cheap right thing here right away. If somebody forgets to call Start() it's much easier to debug than flakily invoking a half-constructed object.

Reply:
OK, sounds reasonable.  Thanks for the explanation.